src/iosource/ManagerBase.cc

```diff
@@ -59,127 +104,69 @@ IOSource* Manager::FindSoonest(double* ts)
 			break;
 			}
 
-	// Ideally, we would always call select on the fds to see which
-	// are ready, and return the soonest. Unfortunately, that'd mean
-	// one select-call per packet, which we can't afford in high-volume
-	// environments.  Thus, we call select only every SELECT_FREQUENCY
-	// call (or if all sources report that they are dry).
+	// If there aren't any sources and exit_only_after_terminate is false, just
+	// return an empty set of sources. We want the main loop to end.
+	if ( Size() == 0 && ( ! BifConst::exit_only_after_terminate || terminating ) )
+		return;
 
-	++call_count;
+	double timeout = -1;
+	IOSource* timeout_src = nullptr;
 
-	IOSource* soonest_src = 0;
-	double soonest_ts = 1e20;
-	double soonest_local_network_time = 1e20;
-	bool all_idle = true;
+	++poll_counter;
+	bool time_to_poll = (poll_counter & POLL_FREQUENCY) == 0;
 
-	// Find soonest source of those which tell us they have something to
-	// process.
-	for ( SourceList::iterator i = sources.begin(); i != sources.end(); ++i )
+	// Find the source with the next timeout value.
+	for ( auto src : sources )
 		{
-		if ( ! (*i)->src->IsIdle() )
+		if ( src->src->IsOpen() )
 			{
-			all_idle = false;
-			double local_network_time = 0;
-			double ts = (*i)->src->NextTimestamp(&local_network_time);
-			if ( ts >= 0 && ts < soonest_ts )
+			double next = src->src->GetNextTimeout();
+			if ( timeout == -1 || ( next >= 0.0 && next < timeout ) )
 				{
-				soonest_ts = ts;
-				soonest_src = (*i)->src;
-				soonest_local_network_time =
-					local_network_time ?
-						local_network_time : ts;
+				timeout = next;
+				timeout_src = src->src;
+
+				// If we found a source with a zero timeout, just return it immediately.
+				// This is a fast optimization, but we want to make sure to go through
+				// with the poll periodically to avoid starvation.
+				if ( timeout == 0 && zero_timeout_count % 100 != 0 )
+					{
+					zero_timeout_count++;
+					ready->push_back(timeout_src);
+					return;
```

Comment:
Think there is a hypothetical starvation issue with this logic:

If there is a source with timeout of 0 exactly every `POLL_FREQUENCY` iterations then this early-return will bypass the `Poll`, but then also every other iteration could bypass the `Poll` when there's a live packet source (and in that case the `zero_timeout_count % 100` throttling also gets bypassed since `zero_timeout_count` gets set outside the loop).

Think I would remove the early-return here and just always iterate all source (i.e. maybe multiple sources with timeout 0, and those should just get processed together).  Can maybe even get rid of the `zero_timeout_count` throttling if `POLL_FREQUENCY` can now cover the starvation issue.  E.g.:

```
	if ( ready->empty() || time_to_poll )
		Poll(ready, timeout, timeout_src);
```

Reply:
The problem with removing the return is that now you might pass `ready` with things it to `Poll`, which then might insert duplicates into the vector. `ready` was a `std::set` way back when to avoid just this problem but `set`'s insert is too slow for this usage.